### DISCUSSION TITLE
Return to old crypto lib for Element-IOS instead of disabling Share Extension

### DISCUSSION THREAD

**Discussion by jacotec** [element-hq/element-meta#1897] [Ideas]
As we know meanwhile, the Rust crypto lib has a major issue being not multitask-safe. Designing it like this for multitasking environments was really an epic fail, but shit happens and nobody's hurt ;-)

However, to work around it until Element-X is completed with all features (which will probably take a few months), **in my opinion it is really the wrong way to disable the Share Extension for millions of users**!

Sharing is a basic feature in IOS, was supported since years in Element-IOS and is highly used. Even more, any other messaging app supports it!

It will be hard to explain it to our users, after working really hard to get them away from mainstream walled-garden-messengers and to Matrix/Element, that they will loose a basic feature and need to work around this multiple times per day and for the next months.

We'll loose many users with this.

I fully understand that the Element team won't spend a week or more to fix the issue in an app which days are already counted, but as long as we don't have a fully working replacement, **the only solution in my opinion is to revert to the old crypto lib**, which worked pretty well and did not have this issue - and switch back to the Rust crypto in a fixed version with EX.

**Comment by manuroe** [comment]
It is impossible to go back to the old crypto. The migration is a multi-month work. 
Having a crypto implementation per platform were too much work to maintain. It meant x3 more work for each new feature or security issue. The new rust crypto lib is more stable. We see less Unable to Decrypt errors (UTDs) since it has been adopted on iOS and Android. If there is a problem in the crypto, we will be quicker to fix it.

It has been a hard choice but we decided that we should shut down features that do not work well enough or break on classic element apps so that we can focus on Element X apps. We want to release EX as soon as possible and to put maximal efforts on it.
We are aware about the UX consequences but we just cannot be on several fronts.

**Reply by jacotec** [reply]
Thanks for your clarification. One thing I don't understand: Why is the choice between the old crypto and the Rust crypto so hard? Not long ago there was a switch in the settings - can't this switch be simply brought back?

**Reply by manuroe** [reply]
The setting was one-way. It was not possible to come back to the "old" crypto.
It is not easy to switch the crypto libs. You have to migrate the keys from one store to another. The migration is available only in one direction.

**Reply by jacotec** [reply]
OK, thanks for the insight 😊

So, I'll keep my thumbs pressed that EX goes forward. OIDC and Voice Messages and I can start testing it 😏
